README

abstract:
  wdfs is a webdav filesystem with special features for accessing subversion
  repositories. it is based on fuse v2.5+ and neon v0.24.7+.


author of wdfs:
  (c) 2005 - 2007 jens m. noedler, noedler@web.de


license:
  wdfs is distributed under the terms of the gnu general public license.
  see file COPYING for details.


home and svn repository of wdfs:
  http://noedler.de/projekte/wdfs/
  http[s]://svn.projectai.org/svn/projects_private/wdfs/


features:
 - generic webdav filesystem
 - http redirect support
 - https support
 - file locking support (different modes)
 - access to all revisions of a webdav exported subversion repository
 - versioning filesystem for autoversioning enabled subversion repositories
   (see section "wdfs, subversion and apache" in this document)


dependencies:
 - operating systems: linux (kernel 2.4 or 2.6), freebsd (6.x or 7.x), mac os x
   hint for linux user: using a kernel 2.6.15 or later is recommended.
 - fuse (filesystem in userspace) v2.5 or later (http://fuse.sourceforge.net/)
 - neon webdav library v0.24.7 or later (http://www.webdav.org/neon/)


installation:
  $ tar xfz wdfs-1.x.x.tar.gz
  $ cd wdfs-1.x.x
  $ ./configure && make
  $ su -c "make install"


usage:
  $ mkdir ~/mountpoint
  $ wdfs http[s]://server[:port][/directory/] ~/mountpoint [options]
  $ wdfs -h      # prints wdfs and fuse options


mounting wdfs using fstab:
  example entry for the /etc/fstab file:
  wdfs#http://server  /mnt/wdfs  fuse  users,locking=simple  0  0

  hint: due to a bug in fuse 2.6.3 and earlier it's not possible to add
  the option "username=abc" to the option list. this will be fixed with 
  later versions of fuse.


wdfs, subversion and apache:
  wdfs may be used to access subversion repositories that are exported as
  webdav using the apache webserver. you need to load the required apache 
  modules for webdav and for subversion and add a location directive like
  the following.

  <Location "/svn">
    DAV svn
    SVNPath /home/jens/repositories/wdfs
    SVNAutoversioning on
    AuthType None
  </Location>


version number conventions:
 - the major number will never change.
 - the minor number changes if new features are introduced or 
   if parts of the api are changed.
   warning: this may include backward-incompatible changes!
 - the last number changes on bug fixes and little non-api changes.
   e.g. v1.1.0 introduced new features from it's predecessor v1.0.1


limitations of this implementation:
 - wdfs is not (yet) multithread safe. it always uses fuse single thread mode.
 - read/write of big files (~ 50 MB+) is slow, because the _complete_ file
   will be GET/PUT on a request.
 - svn mode: only up to 2^31-1 revisions can be accessed,
   because "latest_revision" is an integer variable.
 - svn mode: if a subdirectory of a subversion repository is mounted, the
   "svn_basedir" shows the content of the repository root and not only the
   content of the mounted subdir. [caused by svn_get_remotepath()]
 - http redirect support:
    * there is no redirect support for the mountpoint. please mount the new
      location, that will be printed as an error message.
    * the redirect support is limited to the webdav server, that is mounted.
      redirects to another server are not supported.


limitations of wdfs, caused by other limitations:
 - wdfs supports only utf-8 encoded file and directory names for svn access.
   use utf8 capable applications or pure ascii (no iso8859!) encoded names.
   (limitation caused by subversion, which stores all data utf-8 encoded)
 - wdfs does not support setting file attributes with utime() or utimes().
   (e.g. used by "touch file -m -t 200601010000"). see wdfs_setattr() for
   details. (limitation caused by webdav specification rfc 2518)


source code conventions:
 - variable name "remotepath" is used for WebDAV access (/remote/dir/file)
 - variable name "localpath" is used for local access (/file)
 - variable name "ret" is used for the return values of methods
 - methods and structs starting with "ne_" are part of the neon library
 - error output:
    - "## " prefix for warnings and error messages
 - debug output:
    - ">> " prefix for fuse callback methods
    - "** " prefix for cache related messages (no errors)
    - "++ " prefix for locking related messages (no errors)
 - the code is formated with tabs (tab width 4 characters)
 - every line should not be longer than 80 characters


acknowledgements:
  helmut neukirchen for supervising my bachelor thesis and the development
  of wdfs. miklos szeredi for developing fuse and being a great maintainer.
  the members of the fuse-devel mailing list for helping me and everybody 
  who is contributing to wdfs. thanks!




ChangeLog
2007-04-17  noedler

	wdfs, 1.4.2

2007-04-17  noedler

	- fixed a stupid bug that avoids mounting the root of a server
	- documented the optional port for mounting a webdav server
	- improved attribute caching for some special cases

2007-02-05  noedler

	- wdfs, 1.4.1
	- updated README
	- added redirect handling for rename() and unlink(). thanks joseph!

2007-01-25  noedler

	compile fix for mac os x: define strndup() for apple computers

2007-01-24  noedler

	fixed mounting a webdav share that redirects (e.g.) from '/foo' to '/foo/'. thanks to joseph dunn for the patch.

2007-01-19  noedler

	wdfs, 1.4.0

2007-01-14  noedler

	enable http redirect support by default and add option "no_redirect" to disable it

2007-01-14  noedler

	coding style...

2007-01-13  noedler

	- updated copyright to 2007
	- added dummy impl of wdfs_init()
	- updated README with an example how to mount wdfs thru fstab

2007-01-13  noedler

	updated README file

2006-11-13  noedler

	added a dummy implementation for statfs()

2006-11-09  noedler

	fixed a bug in the assert statements, thanks sven bachmann for pointing it out!

2006-10-27  noedler

	fixed creating the ChangeLog

2006-10-27  noedler

	Initial commit for the public SVN

2006-10-17  jens m. noedler  <noedler@web.de>

	- introduced unify_path() which does the escaping/unescaping
	  the remotepath in one place
	- added checks for the return value of remove_ending_slash()
	- new: remotepath_basedir may contain special chars
	- some random coding style improvements

2006-10-17  jens m. noedler  <noedler@web.de>

	improvement: return -EPERM instead of -ENOENT if deleting fails
	with "403 Forbidden". patch from joseph.dunn@colorado.edu

2006-10-16  jens m. noedler  <noedler@web.de>

	just some random coding style improvements

2006-10-14  jens m. noedler  <noedler@web.de>

	- fixed parsing of "-o accept_sslcert" parameter
	- added check of "-o locking_timeout" parameter

2006-10-14  jens m. noedler  <noedler@web.de>

	converted most printf() to fprintf(stderr, ...)

2006-10-14  jens m. noedler  <noedler@web.de>

	also allow to enter authentication information (username and password)
	via the terminal, based on a patch from joseph.dunn@colorado.edu. thanks!

2006-10-13  jens m. noedler  <noedler@web.de>

	- new command line parsing using the fuse option functions
	  thanks to yuri vasilevski for the patch!
	- wdfs require due to this change fuse 2.5 or later

2006-08-22  jens m. noedler  <noedler@web.de>

	- replaced "DT_DIR << 12" with "S_IFDIR | 0777" for "." and ".." dentries
	- removed useless includes of time.h

2006-08-21  jens m. noedler  <noedler@web.de>

	bugfix: set . and .. to type DT_DIR
	thanks for the patch from mathieu boespflug

2006-08-01  jens m. noedler  <noedler@web.de>

	wdfs, 1.3.1

2006-08-01  jens m. noedler  <noedler@web.de>

	added these statement to permit debian to add wdfs to it's archive

	 *  This program is released under the GPL with the additional exemption
	 *  that compiling, linking and/or using OpenSSL is allowed.

	details: http://www.gnome.org/~markmc/openssl-and-the-gpl.html

2006-07-26  jens m. noedler  <noedler@web.de>

	wdfs, 1.3.0

2006-07-26  jens m. noedler  <noedler@web.de>

	implemented dummy wdfs_chmod() method

2006-07-26  jens m. noedler  <noedler@web.de>

	fixed some escape/unescape problems

2006-07-26  jens m. noedler  <noedler@web.de>

	fix: invalid read in remove_ending_slash()

2006-07-25  jens m. noedler  <noedler@web.de>

	simplified #definition of NEON_VERSION

2006-07-25  jens m. noedler  <noedler@web.de>

	- do not allow lock_timeout to be set to 0 seconds
	- fixed a bug regarding too late unescaping of the remotepath,
	  what may lead to recursive adding of the directory to itself.

2006-07-25  jens m. noedler  <noedler@web.de>

	memory management: replaced wdfs_free() through FREE() and free_chars()

2006-07-25  jens m. noedler  <noedler@web.de>

2006-07-24  jens m. noedler  <noedler@web.de>

	wdfs is now compatible to neon 0.26, thanks again to suse

2006-07-21  jens m. noedler  <noedler@web.de>

	type updates and conversion of error_parameter_parsing from int to bool_t

2006-07-21  jens m. noedler  <noedler@web.de>

	correct indentation and fix for "the last parameter parsing bug" ;-)

2006-07-20  jens m. noedler  <noedler@web.de>

	typo updates

2006-07-20  jens m. noedler  <noedler@web.de>

	conversion from NE_FREE(...) to wdfs_free(...)

2006-07-20  jens m. noedler  <noedler@web.de>

	enhanced argument handling, also based on suse patches

2006-07-20  jens m. noedler  <noedler@web.de>

	reset arguments to avoid storing sensitive data in the process table
	based on a patch from max@suse.de

2006-07-20  jens m. noedler  <noedler@web.de>

	fix for 64 bit platforms due to missing braces
	patch from max@suse.de, thanks!

2006-06-27  jens m. noedler  <noedler@web.de>

	added a comment about freebsd compat in the README

2006-06-27  jens m. noedler  <noedler@web.de>

	wdfs, 1.2.1
	just a build fix for freebsd

2006-06-26  jens m. noedler  <noedler@web.de>

	compile fix for freebsd

2006-04-10  jens m. noedler  <noedler@web.de>

	wdfs, 1.2.0
	updated the README

2006-04-10  jens m. noedler  <noedler@web.de>

	finally fixed the last ftruncate() bug
	thanks to miklos! :-)

2006-04-06  jens m. noedler  <noedler@web.de>

	ftruncate() done.
	but there is still a bug with ftruncate() and O_APPEND

2006-03-12  jens m. noedler  <noedler@web.de>

	made redirect support configurable via parameter "-r"
	("-r" is no longer used by fuse, I think since v2.5)

2006-03-12  jens m. noedler  <noedler@web.de>

	smoother redirect support using
	static int handle_redirect(char **remotepath);

2006-03-12  jens m. noedler  <noedler@web.de>

	stub for wdfs_ftruncate

2006-03-12  jens m. noedler  <noedler@web.de>

	some changes for FUSE_USE_VERSION 25
	and compat with fuse 2.3 and 2.4

2006-03-11  jens m. noedler  <noedler@web.de>

	first stage of redirect support
	first steps to go to fuse interface v25


src/wdfs-main.h
#ifndef WDFSMAIN_H_
#define WDFSMAIN_H_

#ifdef HAVE_CONFIG_H
	#include <config.h>
#endif

#define FUSE_USE_VERSION 25

#include <fuse.h>
#include <ne_basic.h>

/* build the neon version, which is not directly exported by the neon library */
#if defined(NE_FEATURE_TS_SSL)	/* true for neon 0.26+  */
	#define NEON_VERSION 26
#elif defined(NE_FEATURE_SSL)	/* true for neon 0.25+  */
	#define NEON_VERSION 25
#else							/* neon 0.24 is the minimal requirement */
	#define NEON_VERSION 24
#endif
/* 	it's also possible to replace the above with the following: 
	(file configure.ac, after the PKG_CHECK_MODULES call)

	case `pkg-config --modversion neon` in
		0.24*) AC_DEFINE(NEON_VERSION, 24,
				[The minor version number of the neon library]) ;;
		0.25*) AC_DEFINE(NEON_VERSION, 25) ;;
		*)     AC_DEFINE(NEON_VERSION, 26) ;;
	esac
*/

typedef enum {
	true 	= 1,
	false 	= 0
} bool_t;

/* used as mode for unify_path() */
enum {
	ESCAPE     = 0x0,
	UNESCAPE   = 0x1,
	/* do not remove trailing slashes */
	LEAVESLASH = 0x2
};

struct wdfs_conf {
	/* the name of the wdfs executable */
	char *program_name;
	/* if set to "true" wdfs specific debug output is generated */
	bool_t debug;
	/* if set to "true" every certificate is accepted without asking the user */
	bool_t accept_certificate;
	/* username of the webdav resource */
	char *username;
	/* password of the webdav resource */
	char *password;
	/* if set to "true" enables http redirect support */
	bool_t redirect;
	/* if set to "true" enables transparent access to all svn revisions in
	 * a repository thru a virtual directory. */
	bool_t svn_mode;
	/* locking mode of files */
	int locking_mode;
	/* timeout for a lock in seconds */
	int locking_timeout;
	/* address of the webdav resource we are connecting to */
	char *webdav_resource;
};

extern struct wdfs_conf wdfs;

/* look at wdfs-main.c for comments on these extern variables */
extern const char *project_name;
extern char *remotepath_basedir;

/* used by wdfs_readdir() and by svn.h/svn.c to add files to requested 
 * directories using fuse's filler() method. */
struct dir_item {
	void *buf;
	fuse_fill_dir_t filler;
	char *remotepath;
};

char* remove_ending_slashes(const char *in);
char* unify_path(const char *in, int mode);
void free_chars(char **arg, ...);

/* takes an lvalue and sets it to NULL after freeing. taken from neon. */
#define FREE(x) do { if ((x) != NULL) free((x)); (x) = NULL; } while (0)

#ifndef HAVE_STRNDUP
	char* strndup(const char *s, size_t n);
#endif

#endif /*WDFSMAIN_H_*/
